### PR TITLE
Fix wrong mode being pickled for the initial state only

### PR DESCRIPTION
**Pull request by NeatMonster**
If the entrypoint of an program is located on a Thumb mode instruction (if the
least significant bit of the entrypoint is set), the wrong mode will be saved in
the pickled initial state because the mode switching is only happening later.

This commit fixes this issue by switching mode before the pickling, and
unsetting the LSbit of the PC to avoid reading the wrong instruction bytes.

---
This change is [Reviewable](https://reviewable.io/reviews/trailofbits/manticore/1241)

### CODE REVIEW

File: manticore/platforms/linux.py [hunk 1]
```diff
@@ -2413,6 +2413,8 @@ def _arch_specific_init(self):
         elif self.arch == 'armv7':
             self._uname_machine = 'armv71'
             self._init_arm_kernel_helpers()
+            self.current._set_mode_by_val(self.current.PC)
+            self.current.PC &= ~1
```
Comment (disconnect3d) [inline manticore/platforms/linux.py]: Is this correct? Shouldn't the lsb of pc be ignored anyway?

**Comment by NeatMonster** [comment]
When loading an ELF binary having its entrypoint on a Thumb instruction (aka if the least significant bit of the entrypoint is set), `cpu.PC` will also end up with the LSB set. The execution will then fail.

Please consider the following example:
```
from manticore.core.cpu.arm import Armv7Cpu as Cpu
from manticore.core.memory import Memory32

from capstone import *
from keystone import *

#####
import logging
from manticore.utils.log import all_loggers

for logger_name in all_loggers:
    logger = logging.getLogger(logger_name)
    if logger_name == "manticore.manticore" \
            or logger_name.startswith("manticore.core.smtlib"):
        logger.setLevel(logging.INFO)
    else:
        logger.setLevel(logging.DEBUG)
#####

ks = Ks(KS_ARCH_ARM, KS_MODE_THUMB)


def assemble(asm):
    ords = ks.asm(asm)[0]
    if not ords:
        raise Exception(f'bad assembly: {asm}')
    return ''.join(map(chr, ords))


def setup_cpu(cpu, asm):
    code = cpu.memory.mmap(0x1000, 0x1000, 'rwx')
    data = cpu.memory.mmap(0xd000, 0x1000, 'rw')
    stack = cpu.memory.mmap(0xf000, 0x1000, 'rw')
    cpu.memory.write(code, assemble(asm))
    cpu.regfile.write('PC', code)
    cpu.regfile.write('SP', stack + 0x1000)

def test(func, name):
    cpu = Cpu(Memory32())
    mem = cpu.memory
    reg = cpu.regfile

    asm = 'add r1, r2, r3'
    setup_cpu(cpu, asm)

    func(cpu)

    reg.write('R1', 0)
    reg.write('R2', 0x1234)
    reg.write('R3', 0x5678)
    cpu.execute()
    print(name, "succeeded" if reg.read('R1') == 0x68ac else "failed")

def set_pc(cpu):
    cpu.PC = 0x1001

def set_pc_fix(cpu):
    cpu.PC = 0x1001
    cpu._set_mode_by_val(cpu.PC)
    cpu.PC &= ~1

test(set_pc, 'pc')
test(set_pc_fix, 'pc fix')
```

and corresponding output:
```
2018-11-01 13:05:25,810: [28177] m.c.memory:DEBUG: New memory map @1000 size:1000
2018-11-01 13:05:25,810: [28177] m.c.memory:DEBUG: New memory map @d000 size:1000
2018-11-01 13:05:25,810: [28177] m.c.memory:DEBUG: New memory map @f000 size:1000
2018-11-01 13:05:25,811: [28177] m.c.c.abstractcpu:DEBUG: INSTRUCTION: 0x0000000000001001:	andeq	r0, r1, fp, ror #7
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: APSR: 0x0000000000000000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R0 : 0x0000000000000000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R1 : 0x0000000000000000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R10: 0x0000000000000000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R11: 0x0000000000000000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R12: 0x0000000000000000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R13: 0x0000000000010000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R14: 0x0000000000000000
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R15: 0x0000000000001001
2018-11-01 13:05:25,811: [28177] m.c.c.a.registers:DEBUG: R2 : 0x0000000000001234
2018-11-01 13:05:25,812: [28177] m.c.c.a.registers:DEBUG: R3 : 0x0000000000005678
2018-11-01 13:05:25,812: [28177] m.c.c.a.registers:DEBUG: R4 : 0x0000000000000000
2018-11-01 13:05:25,812: [28177] m.c.c.a.registers:DEBUG: R5 : 0x0000000000000000
2018-11-01 13:05:25,812: [28177] m.c.c.a.registers:DEBUG: R6 : 0x0000000000000000
2018-11-01 13:05:25,812: [28177] m.c.c.a.registers:DEBUG: R7 : 0x0000000000000000
2018-11-01 13:05:25,812: [28177] m.c.c.a.registers:DEBUG: R8 : 0x0000000000000000
2018-11-01 13:05:25,812: [28177] m.c.c.a.registers:DEBUG: R9 : 0x0000000000000000
pc failed
2018-11-01 13:05:25,812: [28177] m.c.memory:DEBUG: New memory map @1000 size:1000
2018-11-01 13:05:25,812: [28177] m.c.memory:DEBUG: New memory map @d000 size:1000
2018-11-01 13:05:25,813: [28177] m.c.memory:DEBUG: New memory map @f000 size:1000
2018-11-01 13:05:25,813: [28177] m.c.c.arm:DEBUG: swapping into THUMB mode
2018-11-01 13:05:25,813: [28177] m.c.c.abstractcpu:DEBUG: INSTRUCTION: 0x0000000000001000:	add.w	r1, r2, r3
2018-11-01 13:05:25,813: [28177] m.c.c.a.registers:DEBUG: APSR: 0x0000000000000000
2018-11-01 13:05:25,813: [28177] m.c.c.a.registers:DEBUG: R0 : 0x0000000000000000
2018-11-01 13:05:25,813: [28177] m.c.c.a.registers:DEBUG: R1 : 0x0000000000000000
2018-11-01 13:05:25,813: [28177] m.c.c.a.registers:DEBUG: R10: 0x0000000000000000
2018-11-01 13:05:25,813: [28177] m.c.c.a.registers:DEBUG: R11: 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R12: 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R13: 0x0000000000010000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R14: 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R15: 0x0000000000001000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R2 : 0x0000000000001234
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R3 : 0x0000000000005678
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R4 : 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R5 : 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R6 : 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R7 : 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R8 : 0x0000000000000000
2018-11-01 13:05:25,814: [28177] m.c.c.a.registers:DEBUG: R9 : 0x0000000000000000
pc fix succeeded
```

**Comment by disconnect3d** [comment]
Can you add it as a test and we will merge it? :)

**Comment by NeatMonster** [comment]
Here is an unit test for it. It ensures that if the entrypoint is on a Thumb mode instruction, the instruction will be fetched correctly and executed in the right mode. I had to craft a special binary as the issue is with the Linux loader.

File: tests/test_linux.py [hunk 2]
```diff
@@ -274,3 +274,21 @@ def test_serialize_state_with_closed_files(self):
...
+    def test_thumb_mode_entrypoint(self):
+        m = Manticore.linux(os.path.join(os.path.dirname(__file__), 'binaries', 'thumb_mode_entrypoint'))
+
+        @m.init
+        def init(state):
+            state.cpu.regfile.write('R0', 0)
+            state.cpu.regfile.write('R1', 0x1234)
+            state.cpu.regfile.write('R2', 0x5678)
+
+            @m.hook(None)
```
Comment (disconnect3d) [inline tests/test_linux.py]: Is this indentation here proper? Shouldn't it be one less level?

Comment (disconnect3d) [reply tests/test_linux.py]: also: is it good to assume the hook will fire? Maybe we should put a:

```python
hook_fired = False

@m.hook(None)
def success(state):
    # ...
    hook_fired = True

self.assertTrue(hook_fired)
```
?

**Comment by disconnect3d** [comment]
Hmm it seems the tested code crashes for me, any idea why?

For a `testme.py` file:
```python
from manticore import Manticore

m = Manticore.linux('./thumb_mode_entrypoint')
m.verbosity(5)

@m.init
def init(state):
    state.cpu.regfile.write('R0', 0)
    state.cpu.regfile.write('R1', 0x1234)
    state.cpu.regfile.write('R2', 0x5678)

@m.hook(None)
def success(state):
    if state.cpu.PC < 0x1004:
        return
    self.assertEqual(state.cpu.regfile.read('R0'), 0x68ac)
    state.abandon()

m.run()
```
launching it segfaults:
```
$ python testme.py
Exception ignored in: <bound method FileMap.__del__ of <FileMap [./thumb_mode_entrypoint+74] 0x0000000000001000-0x0000000000002000 r x>>
Traceback (most recent call last):
  File "/home/dc/venvs/manticore/lib/python3.6/site-packages/manticore-0.2.2-py3.6.egg/manticore/core/memory.py", line 331, in __del__
    munmap(self._data, self._mapped_size)
  File "/home/dc/venvs/manticore/lib/python3.6/site-packages/manticore-0.2.2-py3.6.egg/manticore/utils/mappings.py", line 55, in munmap
    assert result == 0
AssertionError:
Exception ignored in: <bound method FileMap.__del__ of <FileMap [./thumb_mode_entrypoint+78] 0x0000000000002000-0x0000000000003000 rw >>
Traceback (most recent call last):
  File "/home/dc/venvs/manticore/lib/python3.6/site-packages/manticore-0.2.2-py3.6.egg/manticore/core/memory.py", line 331, in __del__
    munmap(self._data, self._mapped_size)
  File "/home/dc/venvs/manticore/lib/python3.6/site-packages/manticore-0.2.2-py3.6.egg/manticore/utils/mappings.py", line 55, in munmap
    assert result == 0
AssertionError:
2018-11-02 16:31:01,702: [3743] m.manticore:DEBUG: Starting 1 processes.
2018-11-02 16:31:01,706: [3743] m.c.s.solver:DEBUG: >(set-logic QF_AUFBV)
2018-11-02 16:31:01,706: [3743] m.c.s.solver:DEBUG: >(set-option :global-decls false)
2018-11-02 16:31:01,706: [3743] m.c.s.solver:DEBUG: >(get-info :version)
2018-11-02 16:31:01,707: [3743] m.c.s.solver:DEBUG: <(:version "4.8.0")
2018-11-02 16:31:01,708: [3743] m.c.s.solver:DEBUG: Z3 version: Version(major=4, minor=8, patch=0)
2018-11-02 16:31:01,714: [3743] m.c.executor:INFO: load state 0
Segmentation fault (core dumped)
```

File: tests/test_linux.py (hunk 2 above)
Comment (NeatMonster) [reply tests/test_linux.py]: Yes, the indentation is as intended. But in retrospect I agree that it maybe isn't the easier to read. I will refactor the unit test a bit in my next commit.

**Comment by NeatMonster** [comment]
I've refactored the unit test to remove the nested hooks, and added some comments.

File: tests/test_linux.py [hunk 3]
```diff
@@ -274,3 +274,33 @@ def test_serialize_state_with_closed_files(self):
         fd = platform.sys_open(filename, os.O_RDONLY, 0o600)
         platform.sys_close(fd)
         pickle.dumps(platform)
+
+    def test_thumb_mode_entrypoint(self):
+        # thumb_mode_entrypoint is a binary with only one instruction
+        #   0x1000: add.w   r0, r1, r2
+        # which is a Thumb instruction, so the entrypoint is set to 0x1001
+        m = Manticore.linux(os.path.join(os.path.dirname(__file__), 'binaries', 'thumb_mode_entrypoint'))
+        m.success = True
```
Comment (disconnect3d) [inline tests/test_linux.py]: Wouldn't it be better to initialize this with `False` too? So if hooks doesn't fire for some reason, we will fail the test.

Comment (NeatMonster) [reply tests/test_linux.py]: Done in my latest commit.

**Pull request merged**